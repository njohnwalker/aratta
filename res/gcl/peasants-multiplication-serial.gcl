// serial version of peasant's multiplication

@req(x >= 0 & y >= 0)
a, b, res := x, y, 0 ;

// invariant a * b + res == x * y

do @inv()
[] a > 0 ->
     if
         a % 2 == 1 -> res := res + b ;
     fi
     a := a / 2 ;
     b := b * 2 ;
od
@ens(res == x * y)
