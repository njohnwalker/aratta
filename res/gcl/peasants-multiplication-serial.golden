@req((0 <= x) & (0 <= y))
a, b, res := x, y, 0 ;
do @inv()
[] a > 0 ->
      if (a % 2) == 1 ->
            res := res + b ;
      fi
      a := a / 2 ;
      b := b * 2 ;
od
@ens(res == (x * y))
